core/main/src/firebolt/handlers/provider_registrar.rs

```diff
@@ -103,138 +132,610 @@ impl ProviderRegistrar {
-                                request_method,
-                                move |params, context| async move {
-                                    let mut params_sequence = params.sequence();
-                                    let call_context: CallContext = params_sequence.next().unwrap();
-                                    let request: ListenRequest = params_sequence.next().unwrap();
-                                    let listening = request.listen;
-
-                                    ProviderBroker::register_or_unregister_provider(
-                                        &context.platform_state,
-                                        attributes.capability.into(),
-                                        attributes.method.into(),
-                                        attributes.event,
-                                        call_context,
-                                        request,
-                                    )
-                                    .await;
-
-                                    Ok(ListenerResponse {
-                                        listening,
-                                        event: attributes.event.into(),
-                                    })
-                                },
-                            )
-                            .unwrap();
-                    }
+    fn register_method(
+        method_name: &'static str,
+        method_type: MethodType,
+        rpc_module: &mut RpcModule<RpcModuleContext>,
+    ) -> bool {
+        info!(
+            "register_method: method_name={}, method_type={:?}",
+            method_name, method_type
+        );
+
+        let result = match method_type {
+            MethodType::AppEventEmitter => {
+                rpc_module.register_async_method(method_name, Self::callback_app_event_emitter)
+            }
+            MethodType::AppEventListener => {
+                rpc_module.register_async_method(method_name, Self::callback_app_event_listener)
+            }
+            MethodType::Error => {
+                rpc_module.register_async_method(method_name, Self::callback_error)
+            }
+            MethodType::Focus => {
+                rpc_module.register_async_method(method_name, Self::callback_focus)
+            }
+            MethodType::Provider => {
+                rpc_module.register_async_method(method_name, Self::callback_register_provider)
+            }
+            MethodType::ProviderInvoker => {
+                rpc_module.register_async_method(method_name, Self::callback_provider_invoker)
+            }
+            MethodType::Response => {
+                rpc_module.register_async_method(method_name, Self::callback_response)
+            }
+        };
+
+        match result {
+            Ok(_) => true,
+            Err(e) => {
+                error!("register_method: Error registering method: {:?}", e);
+                false
+            }
+        }
+    }
+
```

Comment:
Various callback types follow (`callback_*()`), these are effectively executed when the associated firebolt API call is made.